File: server/utils/transcript-filter.ts
interface TranscriptSegment {
    speaker: string
    content: string
    timestamp?: string
}

/**
 * Parse a transcript string into segments with speaker labels
 */
export function parseTranscript(transcript: string): TranscriptSegment[] {
    const segments: TranscriptSegment[] = []
    // Match speaker labels like "SPEAKER_00:", "SPEAKER_01:", "Character1:", etc.
    const speakerPattern = /^([A-Z_0-9]+|Character\d+):\s*(.*)$/gm
    const lines = transcript.split('\n')
    
    let currentSpeaker: string | null = null
    let currentContent: string[] = []
    
    for (const line of lines) {
        const trimmed = line.trim()
        if (!trimmed) continue
        
        const match = trimmed.match(/^([A-Z_0-9]+|Character\d+):\s*(.*)$/)
        if (match) {
            // Save previous segment if exists
            if (currentSpeaker !== null && currentContent.length > 0) {
                segments.push({
                    speaker: currentSpeaker,
                    content: currentContent.join(' ').trim()
                })
            }
            // Start new segment
            currentSpeaker = match[1]!
            currentContent = match[2] ? [match[2]] : []
        } else if (currentSpeaker !== null) {
            // Continue current segment
            currentContent.push(trimmed)
        }
    }
    
    // Add last segment
    if (currentSpeaker !== null && currentContent.length > 0) {
        segments.push({
            speaker: currentSpeaker,
            content: currentContent.join(' ').trim()
        })
    }
    
    return segments
}

/**
 * Highlight matching text in a string
 */
function highlightText(text: string, searchString: string): string {
    if (!searchString || !searchString.trim()) {
        return escapeHtml(text)
    }
    
    const searchLower = searchString.toLowerCase()
    const textLower = text.toLowerCase()
    const parts: string[] = []
    let lastIndex = 0
    let index = textLower.indexOf(searchLower, lastIndex)
    
    // If no matches found, return escaped text
    if (index === -1) {
        return escapeHtml(text)
    }
    
    while (index !== -1) {
        // Add text before match
        if (index > lastIndex) {
            parts.push(escapeHtml(text.substring(lastIndex, index)))
        }
        // Add highlighted match
        parts.push(`<mark class="bg-yellow-200 dark:bg-yellow-900">${escapeHtml(text.substring(index, index + searchString.length))}</mark>`)
        lastIndex = index + searchString.length
        index = textLower.indexOf(searchLower, lastIndex)
    }
    
    // Add remaining text
    if (lastIndex < text.length) {
        parts.push(escapeHtml(text.substring(lastIndex)))
    }
    
    return parts.join('')
}

/**
 * Escape HTML to prevent XSS
 */
function escapeHtml(text: string): string {
    const div = { innerHTML: '' } as any
    div.textContent = text
    return div.innerHTML || text.replace(/[&<>"']/g, (char) => {
        const map: Record<string, string> = {
            '&': '&amp;',
            '<': '&lt;',
            '>': '&gt;',
            '"': '&quot;',
            "'": '&#039;'
        }
        return map[char] || char
    })
}

/**
 * Highlight transcript with search string and/or speaker highlighting
 * Returns highlighted transcript and match count
 */
export function highlightTranscript(
    transcript: string,
    searchString?: string,
    speakers?: string[]
): { highlightedTranscript: string; matchCount: number } {
    const segments = parseTranscript(transcript)
    
    let matchCount = 0
    const lines: string[] = []
    let currentSpeaker: string | null = null
    
    for (const segment of segments) {
        const isSpeakerMatch = speakers && speakers.length > 0 && speakers.some(speaker =>
            segment.speaker === speaker ||
            segment.speaker.toLowerCase() === speaker.toLowerCase() ||
            segment.speaker.toLowerCase().includes(speaker.toLowerCase())
        )
        
        const isContentMatch = searchString && searchString.trim() &&
            segment.content.toLowerCase().includes(searchString.toLowerCase())
        
        const isMatch = isSpeakerMatch || isContentMatch
        
        if (isMatch) {
            matchCount++
        }
        
        // Add speaker label
        if (segment.speaker !== currentSpeaker) {
            if (lines.length > 0) {
                lines.push('')
            }
            const speakerLabel = isSpeakerMatch && speakers && speakers.length > 0
                ? `<mark class="bg-blue-200 dark:bg-blue-900 font-semibold">${escapeHtml(segment.speaker)}:</mark>`
                : `${escapeHtml(segment.speaker)}:`
            lines.push(speakerLabel)
            currentSpeaker = segment.speaker
        }
        
        // Add content with highlighting
        const highlightedContent = searchString && searchString.trim()
            ? highlightText(segment.content, searchString)
            : escapeHtml(segment.content)
        
        lines.push(highlightedContent)
    }
    
    return {
        highlightedTranscript: lines.join('\n'),
        matchCount
    }
}

/**
 * Extract unique speakers from a transcript
 */
export function extractSpeakers(transcript: string): string[] {
    const segments = parseTranscript(transcript)
    const speakers = new Set<string>()
    
    for (const segment of segments) {
        speakers.add(segment.speaker)
    }
    
    return Array.from(speakers).sort()
}

File: backend/utils/transcript_filter.py
"""Transcript filtering and highlighting utilities."""

import html
import re
from typing import Any


def parse_transcript(transcript: str) -> list[dict[str, Any]]:
    """Parse a transcript string into segments with speaker labels."""
    segments: list[dict[str, Any]] = []
    lines = transcript.split('\n')

    current_speaker: str | None = None
    current_timestamp: str | None = None
    current_content: list[str] = []

    # Match: Optional [MM:SS] timestamp, then name followed by colon
    # Supports: "[00:00] Gabe:", "Caroline:", "[12:34] SPEAKER_00:", "John Smith:"
    # Using broad pattern: [A-Z0-9] start, allowed chars, max 60 chars
    speaker_pattern = re.compile(
        r'^(?:\[(\d{1,3}:\d{2})\]\s+)?([A-Z0-9][\w\s\-\'._()]{1,60}?):\s*(.*)$'
    )

    for line in lines:
        trimmed = line.strip()
        if not trimmed:
            continue

        match = speaker_pattern.match(trimmed)
        if match:
            # Save previous segment if exists
            if current_speaker is not None and current_content:
                segment_data: dict[str, Any] = {
                    'speaker': current_speaker,
                    'content': ' '.join(current_content).strip()
                }
                if current_timestamp:
                    segment_data['timestamp'] = current_timestamp
                segments.append(segment_data)
            # Start new segment (group 1=timestamp, group 2=speaker, group 3=content)
            current_timestamp = match.group(1)  # may be None
            current_speaker = match.group(2)
            current_content = [match.group(3)] if match.group(3) else []
        elif current_speaker is not None:
            # Continue current segment
            current_content.append(trimmed)

    # Add last segment
    if current_speaker is not None and current_content:
        segment_data = {
            'speaker': current_speaker,
            'content': ' '.join(current_content).strip()
        }
        if current_timestamp:
            segment_data['timestamp'] = current_timestamp
        segments.append(segment_data)

    return segments


def escape_html(text: str) -> str:
    """Escape HTML to prevent XSS."""
    return html.escape(text)


def highlight_text(text: str, search_string: str) -> str:
    """Highlight matching text in a string (whole word match)."""
    if not search_string or not search_string.strip():
        return escape_html(text)

    pattern = re.compile(r'\b' + re.escape(search_string) + r'\b', re.IGNORECASE)
    parts: list[str] = []
    last_index = 0

    for match in pattern.finditer(text):
        # Add text before match
        if match.start() > last_index:
            parts.append(escape_html(text[last_index:match.start()]))
        # Add highlighted match
        parts.append(
            f'<mark class="bg-yellow-200 dark:bg-yellow-900">'
            f'{escape_html(match.group())}</mark>'
        )
        last_index = match.end()

    # If no matches found, return escaped text
    if not parts:
        return escape_html(text)

    # Add remaining text
    if last_index < len(text):
        parts.append(escape_html(text[last_index:]))

    return ''.join(parts)


def highlight_transcript(
    transcript: str,
    search_string: str | None = None,
) -> dict[str, Any]:
    """
    Highlight search term occurrences in transcript text.

    Returns dict with highlightedTranscript and matchCount (total word matches).
    """
    segments = parse_transcript(transcript)

    match_count = 0
    lines: list[str] = []
    current_speaker: str | None = None
    word_pattern = re.compile(r'\b' + re.escape(search_string) + r'\b', re.IGNORECASE) if search_string and search_string.strip() else None

    for segment in segments:
        # Count individual word matches (not just segments)
        if word_pattern:
            match_count += len(word_pattern.findall(segment['content']))

        # Add speaker label with optional timestamp
        if segment['speaker'] != current_speaker:
            if lines:
                lines.append('')
            timestamp = segment.get('timestamp')
            if timestamp:
                lines.append(f'[{escape_html(timestamp)}] {escape_html(segment["speaker"])}:')
            else:
                lines.append(f'{escape_html(segment["speaker"])}:')
            current_speaker = segment['speaker']

        # Add content with highlighting
        if search_string and search_string.strip():
            highlighted_content = highlight_text(segment['content'], search_string)
        else:
            highlighted_content = escape_html(segment['content'])

        lines.append(highlighted_content)

    return {
        'highlightedTranscript': '\n'.join(lines),
        'matchCount': match_count
    }


def extract_speakers(transcript: str) -> list[str]:
    """Extract unique speakers from a transcript."""
    segments = parse_transcript(transcript)
    speakers = set()

    for segment in segments:
        speakers.add(segment['speaker'])

    return sorted(speakers)


def calculate_speaker_frequencies(
    transcript: str,
    search_string: str
) -> list[dict[str, Any]]:
    """Calculate search term frequency per speaker."""
    if not search_string or not search_string.strip():
        return []

    segments = parse_transcript(transcript)
    frequency_map: dict[str, int] = {}
    search_lower = search_string.lower()

    word_pattern = re.compile(r'\b' + re.escape(search_lower) + r'\b')

    for segment in segments:
        content_lower = segment['content'].lower()
        count = len(word_pattern.findall(content_lower))

        if count > 0:
            current = frequency_map.get(segment['speaker'], 0)
            frequency_map[segment['speaker']] = current + count

    # Convert to list and sort by count descending
    return sorted(
        [{'speaker': speaker, 'count': count} for speaker, count in frequency_map.items()],
        key=lambda x: x['count'],
        reverse=True
    )
